### ISSUE TITLE
Loss Calculation in Image Captioning 

### ISSUE THREAD

**Issue by arijitx** [yunjey/pytorch-tutorial#101]
Hi , am new to PyTorch in model.py you do ,

```
embeddings = self.embed(captions)
embeddings = torch.cat((features.unsqueeze(1), embeddings), 1)
```

that means you are only passing the context(from the EncoderCNN ) as input to the first time step, hence your **outputs** will always be **1+captions size**  as you are running the lstm for captions+1 number of time steps in the decoder , but looks like you have targets and outputs of same size. What am I missing here ? 
Thanks in advance .

**Comment by icfear** [comment]
Every caption starts with `<start>` and ends with `<end>` in the data_loader. If ground truth is `"there is a cat"`, the caption will be `"<start>  there is a cat <end>" `(actually their corresponding ids in the vocabulary, but for simplicity we can ignore that). `collate_fn` in data_loader returns lengths too, so we will have a caption with length 5. Look at the following line;

https://github.com/yunjey/pytorch-tutorial/blob/6c785eb810848f259431865b12bb8795f4abcb23/tutorials/03-advanced/image_captioning/model.py#L52

Each caption will be embedded with their respective length. Packing `"features <start>  there is a cat <end>" ` with length 5 is the same as packing `"features <start>  there is a cat" `. So,  `<end>`  is discarded.

**Comment by RicherMans** [comment]
@okanlv 
But how about the targets then? Shouldn't they be `there is a cat <end>` instead of the current `<start> there is a cat <end>` ? 
Problem is the following line:

```python
targets = pack_padded_sequence(captions, lengths, batch_first=True)[0]
```

**Comment by okanlv** [comment]
@RicherMans I made a mistake in the previous comment. `collate_fn` returns 6, not 5 for "<start>` there is a cat <end>". `"features <start> there is a cat <end>"` has length 7, so `<end>` is discarded during packing. On the other hand, `"<start> there is a cat <end>"` has length 6, so nothing is discarded.

**Comment by RicherMans** [comment]
But shouldn't the decoder input/output scheme be the following:

Input:
`feature <start> there is a cat <end>`

Output:
`there is a cat <end>` 
?

**Comment by okanlv** [comment]
Decoder should have the same number of elements in both its input and output. Input in your example contains 7 words, whereas the output contains 5 words. Every word (including feature) is used to predict the next word, so it is like the following.

Input:
`feature <start> there is a cat`

Output:
`<start> there is a cat <end>`

**Comment by RicherMans** [comment]
@okanlv 
Got it, thanks!

**Comment by fmehralian** [comment]
Hey guys,
The reason behind concatenating the feature vector with the input ([here](https://github.com/yunjey/pytorch-tutorial/blob/57afe85b2c7e6bb92918a73b9a9b6a3394c92951/tutorials/03-advanced/image_captioning/model.py#L38)) is not clear for me. Would you please help me figure out the difference between:
1) LSTM(cat(image_feature, input)) [link](

![](https://miro.medium.com/max/1400/1*cW47gMv3Tq5Ex5itMalRmw.png)
)
2) LSTM(input, hidden=image_feature) [link](

![](https://miro.medium.com/max/1024/1*vzFwXFJOrg6WRGNsYYT6qg.png)
)
